Self:: doesn't work in patterns

```rust
use async_trait::async_trait;

#[async_trait]
pub trait Trait {
    async fn handle(&self);
}

pub enum Enum {
    Variant,
}

#[async_trait]
impl Trait for Enum {
    async fn handle(&self) {
        let Enum::Variant = self; // works
        let Self::Variant = self; // does not work
    }
}
```

```console
error[E0401]: can't use generic parameters from outer function
   --> tests/test.rs:552:17
    |
549 |     impl Trait for Enum {
    |     ---- `Self` type implicitly declared here, by this `impl`
...
552 |             let Self::Variant = self;
    |                 ^^^^^^^^^^^^^
    |                 |
    |                 use of generic parameter from outer function
    |                 use a type here instead
```

--- a/src/receiver.rs
+++ b/src/receiver.rs
@@ -4,8 +4,8 @@ use std::mem;
 use syn::punctuated::Punctuated;
 use syn::visit_mut::{self, VisitMut};
 use syn::{
-    parse_quote, Block, Error, ExprPath, ExprStruct, Ident, Item, Macro, Path, PathArguments,
-    QSelf, Receiver, Signature, Type, TypePath, WherePredicate,
+    parse_quote, Block, Error, ExprPath, ExprStruct, Ident, Item, Macro, PatPath, Path,
+    PathArguments, QSelf, Receiver, Signature, Type, TypePath, WherePredicate,
 };
 
 pub fn has_self_in_sig(sig: &mut Signature) -> bool {
@@ -32,6 +32,11 @@ impl VisitMut for HasSelf {
     fn visit_expr_path_mut(&mut self, expr: &mut ExprPath) {
         self.0 |= expr.path.segments[0].ident == "Self";
         visit_mut::visit_expr_path_mut(self, expr);
+    }
+
+    fn visit_pat_path_mut(&mut self, pat: &mut PatPath) {
+        self.0 |= pat.path.segments[0].ident == "Self";
+        visit_mut::visit_pat_path_mut(self, pat);
     }
 
     fn visit_type_path_mut(&mut self, ty: &mut TypePath) {
@@ -172,6 +177,13 @@ impl VisitMut for ReplaceReceiver {
             self.self_to_expr_path(&mut expr.path);
         }
         visit_mut::visit_expr_struct_mut(self, expr);
+    }
+
+    fn visit_pat_path_mut(&mut self, pat: &mut PatPath) {
+        if pat.qself.is_none() {
+            self.self_to_qself_expr(&mut pat.qself, &mut pat.path);
+        }
+        visit_mut::visit_pat_path_mut(self, pat);
     }
 
     fn visit_item_mut(&mut self, i: &mut Item) {

--- a/tests/test.rs
+++ b/tests/test.rs
@@ -532,3 +532,25 @@ pub mod issue73 {
         }
     }
 }
+
+// https://github.com/dtolnay/async-trait/issues/81
+pub mod issue81 {
+    use async_trait::async_trait;
+
+    #[async_trait]
+    pub trait Trait {
+        async fn handle(&self);
+    }
+
+    pub enum Enum {
+        Variant,
+    }
+
+    #[async_trait]
+    impl Trait for Enum {
+        async fn handle(&self) {
+            let Enum::Variant = self;
+            let Self::Variant = self;
+        }
+    }
+}

